// react-best-gradient-color-picker/dist/cjs/components/GradientControls.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const context_js_1 = require("../context.js");
const formatters_js_1 = require("../utils/formatters.js");
const Controls_js_1 = require("./Controls.js");
const icon_js_1 = __importStar(require("./icon.js"));
const GradientType = () => {
    const { gradientType, internalOnChange, value } = (0, context_js_1.usePicker)();
    const isLinear = gradientType === 'linear-gradient';
    const isRadial = gradientType === 'radial-gradient';
    const handleLinear = () => {
        const remaining = value.split(/,(.+)/)[1];
        internalOnChange(`linear-gradient(90deg, ${remaining}`);
    };
    const handleRadial = () => {
        const remaining = value.split(/,(.+)/)[1];
        internalOnChange(`radial-gradient(circle, ${remaining}`);
    };
    return (react_1.default.createElement("div", { className: "rbgcp-control-btn-wrapper" },
        react_1.default.createElement("div", { onClick: handleLinear, id: "rbgcp-linear-btn", className: `rbgcp-control-btn rbgcp-linear-btn ${isLinear && 'rbgcp-control-btn-selected'}` },
            react_1.default.createElement(icon_js_1.LinearIcon, { color: isLinear ? '#568CF5' : '' })),
        react_1.default.createElement("div", { onClick: handleRadial, id: "rbgcp-gradient-btn", className: `rbgcp-control-btn rbgcp-gradient-btn ${isRadial && 'rbgcp-control-btn-selected'}` },
            react_1.default.createElement(icon_js_1.RadialIcon, { color: isRadial ? '#568CF5' : '' }))));
};
const StopPicker = () => {
    const { currentLeft, handleGradient, currentColor } = (0, context_js_1.usePicker)();
    const handleMove = (newVal) => {
        handleGradient(currentColor, (0, formatters_js_1.formatInputValues)(parseInt(newVal), 0, 100));
    };
    return (react_1.default.createElement("div", { className: "rbgcp-control-btn-wrapper rbgcp-control-input-wrap rbgcp-stop-input-wrap", style: { paddingLeft: 8 } },
        react_1.default.createElement(icon_js_1.StopIcon, null),
        react_1.default.createElement("input", { id: "rbgcp-stop-input", value: currentLeft, className: "rbgcp-control-input rbgcp-stop-input", onChange: (e) => handleMove(e.target.value) })));
};
const DegreePicker = () => {
    const { degrees, internalOnChange, value } = (0, context_js_1.usePicker)();
    const handleDegrees = (e) => {
        const newValue = (0, formatters_js_1.formatInputValues)(e.target.value, 0, 360);
        const remaining = value.split(/,(.+)/)[1];
        internalOnChange(`linear-gradient(${newValue || 0}deg, ${remaining}`);
    };
    return (react_1.default.createElement("div", { className: "rbgcp-control-btn-wrapper rbgcp-control-input-wrap rbgcp-degree-input-wrap" },
        react_1.default.createElement(icon_js_1.DegreesIcon, null),
        react_1.default.createElement("input", { id: "rbgcp-degree-input", value: degrees, className: "rbgcp-control-input rbgcp-degree-input", onChange: (e) => handleDegrees(e) }),
        react_1.default.createElement("div", { style: {
                position: 'absolute',
                right: degrees > 99 ? 0 : degrees < 10 ? 7 : 3,
                top: 1,
                fontWeight: 400,
                fontSize: 13,
            }, className: "rbgcp-degree-icon" }, "\u00B0")));
};
const DeconsteBtn = () => {
    const { deletePoint } = (0, context_js_1.usePicker)();
    return (react_1.default.createElement("div", { onClick: deletePoint, style: { width: 28, ...(0, Controls_js_1.controlBtnStyles)(false) }, id: "rbgcp-point-deconste-btn", className: "rbgcp-control-icon-btn rbgcp-point-deconste-btn" },
        react_1.default.createElement(icon_js_1.default, null)));
};
const GradientControls = ({ hideGradientType, hideGradientAngle, hideGradientStop, }) => {
    const { gradientType } = (0, context_js_1.usePicker)();
    return (react_1.default.createElement("div", { style: {
            marginTop: 12,
            marginBottom: -4,
            paddingLeft: hideGradientType ? 4 : 0,
        }, id: "rbgcp-gradient-controls-wrap", className: "rbgcp-control-btn-wrapper jsb" },
        !hideGradientType && react_1.default.createElement(GradientType, null),
        react_1.default.createElement("div", { style: { width: 53 } }, !hideGradientAngle && gradientType === 'linear-gradient' && (react_1.default.createElement(DegreePicker, null))),
        !hideGradientStop && react_1.default.createElement(StopPicker, null),
        react_1.default.createElement(DeconsteBtn, null)));
};
exports.default = GradientControls;
